File: impls/go/src/step0_repl/step0_repl.go
package main

import (
    "fmt"
    "bufio"
    "os"
    "strings"
)


func inputPrompt(label string) string {
    // NOTE: Figure out why this works
    var s string
    r := bufio.NewReader(os.Stdin)
    for {
        fmt.Fprint(os.Stderr, label)
        s, _ = r.ReadString('\n')
        if s != "" {
            break
        }
    }
    return strings.TrimSpace(s)
}


func READ(x string) string {
    return x
}


func EVAL(x string) string {
    return x
}


func PRINT(x string) string {
    return x
}


func rep(x string) string {
    res_read := READ(x)
    res_eval := EVAL(res_read)
    res_print := PRINT(res_eval)
    return res_print
}


func main() {
    for {
        user_input := inputPrompt("user> ")
        res_rep := rep(user_input)
        fmt.Printf("%v\n", res_rep)
    }
}

File: impls/python/step0_repl.py
import sys


def READ(x):
    return x


def EVAL(x):
    return x


def PRINT(x):
    return x


def rep(x):
    res_read = READ(x)
    res_eval = EVAL(res_read)
    res_print = PRINT(res_eval)
    return res_print


def main():
    while True:
        try:
            user_input = input("user> ")
            res_rep = rep(user_input)
            print(res_rep)
        except EOFError:
            sys.exit("\nbye! :3")


if __name__ == "__main__":
    main()
